File: api-java/src/main/java/fr/huiitre/tools/modules/riot/valorant/application/core/usecase/ValorantAuthService.java
package fr.huiitre.tools.modules.riot.valorant.application.core.usecase;

import fr.huiitre.tools.modules.core.security.infrastructure.EncryptionService;
import fr.huiitre.tools.modules.riot.valorant.application.core.ports.RiotAuthPort;
import fr.huiitre.tools.modules.riot.valorant.application.core.ports.ValorantAuthRepository;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

@Service
public class ValorantAuthService {

    private final RiotAuthPort riotAuthPort;
    private final ValorantAuthRepository valorantAuthRepository;
    private final EncryptionService encryptionService;

    public ValorantAuthService(RiotAuthPort riotAuthPort,
                               ValorantAuthRepository valorantAuthRepository,
                               EncryptionService encryptionService) {
        this.riotAuthPort = riotAuthPort;
        this.valorantAuthRepository = valorantAuthRepository;
        this.encryptionService = encryptionService;
    }

    public String getOrRefreshAccessToken(Long accountId) {
        // 1. Récupération des données chiffrées en base
        ValorantAuthRepository.ValorantAuthData authData = valorantAuthRepository.findById(accountId)
                .orElseThrow(() -> new IllegalArgumentException("RIOT_AUTH_NOT_FOUND"));

        // 2. Déchiffrement du Refresh Token
        String refreshToken = encryptionService.decrypt(authData.encryptedRefreshToken(), authData.iv());

        try {
            // 3. Appel Riot pour un nouvel Access Token
            RiotAuthPort.ValorantAuthResponse riotResponse = riotAuthPort.refresh(refreshToken);

            // 4. Mise à jour du nouveau Refresh Token (Rotation) - pas de changement de pseudo à ce stade
            String newIv = encryptionService.generateIv();
            String newEncryptedRefresh = encryptionService.encrypt(riotResponse.refreshToken(), newIv);

            valorantAuthRepository.save(
                    authData.userId(),
                    riotResponse.puuid(),
                    authData.region(),
                    null,
                    null,
                    null,
                    newEncryptedRefresh,
                    newIv,
                    riotResponse.refreshTokenExpiresAt()
            );

            return riotResponse.accessToken();

        } catch (IllegalArgumentException e) {
            // Si le refresh token est invalide (périmé chez Riot), on nettoie la base
            if ("RIOT_TOKEN_INVALID".equals(e.getMessage())) {
                valorantAuthRepository.deleteById(accountId);
            }
            throw e;
        }
    }

    public long saveAuthData(Long userId, String puuid, String region, String gameName, String tagLine, String label, String refreshToken, LocalDateTime expiresAt) {
        String iv = encryptionService.generateIv();
        String encryptedRefresh = encryptionService.encrypt(refreshToken, iv);
        return valorantAuthRepository.save(userId, puuid, region, gameName, tagLine, label, encryptedRefresh, iv, expiresAt);
    }
}

File: api/Modules/Riot/Valorant/Application/Core/Services/ValorantAuthService.cs
using Tools.Api.Modules.Core.Common.Application.Exceptions;
using Tools.Api.Modules.Riot.Valorant.Application.Core.Ports;

namespace Tools.Api.Modules.Riot.Valorant.Application.Core.Services;

// Détient le refresh token d'un compte lié : il est le seul à le déchiffrer, à le présenter à Riot
// et à ranger celui que Riot renvoie en échange.
//
// Ce n'est pas un use case sécurisé : le renouvellement périodique l'appelle depuis une tâche de
// fond, sans utilisateur authentifié. Ce sont les use cases qui s'en servent qui portent le
// contrôle d'accès.
public sealed class ValorantAuthService(
    IRiotAuthPort riotAuthPort,
    IValorantAuthRepository valorantAuthRepository,
    IValorantTokenCipher tokenCipher)
{
    // Code que l'adaptateur Riot doit poser sur son AppException quand Riot refuse le refresh
    // token lui-même. Il distingue un jeton périmé — le compte est alors délié — d'une panne
    // passagère, qui ne doit rien supprimer.
    public const string TokenInvalidCode = "RIOT_TOKEN_INVALID";

    public async Task<string> GetOrRefreshAccessToken(long accountId)
    {
        var authData = await valorantAuthRepository.FindById(accountId)
            ?? throw AppException.NotFound(
                "RIOT_AUTH_NOT_FOUND",
                "Aucun jeton n'est enregistré pour ce compte Valorant.");

        var refreshToken = tokenCipher.Decrypt(authData.EncryptedRefreshToken, authData.Iv);

        try
        {
            var riotResponse = await riotAuthPort.Refresh(refreshToken);

            // Riot fait tourner le refresh token à chaque échange : ne pas ranger le nouveau
            // rendrait le compte inutilisable au prochain appel.
            var newIv = tokenCipher.GenerateIv();
            var newEncryptedRefresh = tokenCipher.Encrypt(riotResponse.RefreshToken, newIv);

            // Pseudo et libellé restent nuls : l'upsert ne doit pas écraser ce qui est déjà en base.
            await valorantAuthRepository.Save(
                authData.UserId,
                riotResponse.Puuid,
                authData.Region,
                null,
                null,
                null,
                newEncryptedRefresh,
                newIv,
                riotResponse.RefreshTokenExpiresAt);

            return riotResponse.AccessToken;
        }
        catch (AppException exception) when (exception.Code == TokenInvalidCode)
        {
            // Le jeton est mort chez Riot : le garder ne servirait qu'à rejouer l'échec.
            await valorantAuthRepository.DeleteById(accountId);
            throw;
        }
    }

    public Task<long> SaveAuthData(
        long userId,
        string puuid,
        string region,
        string? gameName,
        string? tagLine,
        string? label,
        string refreshToken,
        DateTime expiresAt)
    {
        var iv = tokenCipher.GenerateIv();
        var encryptedRefresh = tokenCipher.Encrypt(refreshToken, iv);

        return valorantAuthRepository.Save(
            userId, puuid, region, gameName, tagLine, label, encryptedRefresh, iv, expiresAt);
    }
}
